src/main/java/org/folio/fqm/service/MigrationService.java

```diff
@@ -1,28 +1,50 @@
 package org.folio.fqm.service;
 
-import java.util.List;
-import org.apache.commons.lang3.NotImplementedException;
+import javax.annotation.CheckForNull;
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.fql.model.Fql;
+import org.folio.fql.service.FqlService;
+import org.folio.fqm.migration.MigratableQueryInformation;
+import org.folio.fqm.migration.MigrationStrategy;
+import org.folio.fqm.migration.MigrationStrategyRepository;
+import org.springframework.beans.factory.annotation.Autowired;
 
+@Log4j2
+@AllArgsConstructor(onConstructor_ = @Autowired)
 public class MigrationService {
 
-  public int getLatestVersion() {
-    // return 1;
-    throw new NotImplementedException();
+  private static final String CURRENT_VERSION = "1";
+
+  private final FqlService fqlService;
+  private final MigrationStrategyRepository migrationStrategyRepository;
+
+  public String getLatestVersion() {
+    return CURRENT_VERSION;
   }
 
-  public boolean isMigrationNeeded(String fqlQuery) {
-    // return true;
-    throw new NotImplementedException();
+  public boolean isMigrationNeeded(@CheckForNull String fqlQuery) {
+    if (fqlQuery == null) {
+      return true;
+    }
+    Fql fql = fqlService.getFql(fqlQuery);
+    return !this.getLatestVersion().equals(fql._version());
   }
 
   public boolean isMigrationNeeded(MigratableQueryInformation migratableQueryInformation) {
     return isMigrationNeeded(migratableQueryInformation.fqlQuery());
   }
 
   public MigratableQueryInformation migrate(MigratableQueryInformation migratableQueryInformation) {
-    // return migratableQueryInformation;
-    throw new NotImplementedException();
-  }
+    while (isMigrationNeeded(migratableQueryInformation)) {
+      for (MigrationStrategy strategy : migrationStrategyRepository.getMigrationStrategies()) {
+        if (strategy.applies(migratableQueryInformation)) {
+          log.info("Applying {} to {}", strategy.getLabel(), migratableQueryInformation);
+          migratableQueryInformation = strategy.apply(migratableQueryInformation);
```

Comment:
Inefficient. Some, particularly algorithms professors, might even say `O(n^2)`.

However, this is not run frequently. Additionally, most of these `.applies` calls should be a simple `.equals`. Something cleverer, such as a `Map` from source versions to `MigrationStrategy` could work, however, there may be cases where we want to perform a migration on a range of versions, or something more strange, so the extensibility is left here.

Additionally, if `migrationStrategyRepository` supplies migrators in the order from V0 -> V∞, this loop should theoretically only need to execute once.